### DISCUSSION TITLE
Plotting Task 3&4

### DISCUSSION THREAD

**Discussion by MajaR01** [Part-II-Computational-Physics/.github#39] [General]
I am not sure how I can extract the values that function gives me and then plot them

**Comment by JNHallen** [comment]
Hi! 

`E, Psi = npl.eigh(CompleteHamiltonian(N, h=h))`

will get the eigenvalues in an array E and the eigenvectors in an array Psi. For the first task you just have to care about the smaller eigenvalue, which will be E[0] (or you can do something like min(E)). Loop over different $N$ and $h$ values and store the smallest eigenvalues in lists. 

One way of doing this would be:

```
hh = np.logspace(-2, 2, 10)
Ns = np.arange(2, 10, 1)

fignumber = 1
for N in Ns:
    GSenergy = []
    for h in hh:
        E, Psi = npl.eigh(CompleteHamiltonian(N, h=h))
        GSenergy.append(E[0])
    
    plt.figure(fignumber)
    # plot hh versus GSenergy

    plt.yscale('symlog')    

    fignumber += 1
```

I would also like it if you try plotting the energy curves for some different $N$ in the same figure. To do that you will need to divide the energies by $N$ (giving you the energy per spin, rather than the total energy).

**Reply by MajaR01** [reply]
I get a number of empty plots

**Reply by JNHallen** [reply]
You need to replace the line `# plot hh versus GSenergy` with some code to actually plot the data. Have a look at [https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.plot](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.plot) to see details of how to do this. 

The most basic line that will do the trick is 
`plt.plot(hh, GSenergy)`
but do try to play around with colours, marker/line styles and labels if you have time.

**Reply by MajaR01** [reply]
Thanks! :)

**Reply by MajaR01** [reply]
What about Task 4? I tried to copy the same code structure, but computing time is exceeded

**Reply by JNHallen** [reply]
For task 4 you should try using the code given in the task:
```
Ns = np.arange(2, 11, 1)
h = 1

timingCompact = []
for N in Ns:
    print(f"N={N}")
    result = %timeit -o np.linalg.eigh(CompleteHamiltonian2(N))
    timingCompact.append(result)
```

The list timingCompact will contain the timeit objects. To plot the average time with standard deviation as errorbar you can use something like:
```
Averages = [time.average for time in timingCompact]
StDev = [time.stdev for time in timingCompact]

plt.errorbar(Ns, Averages, yerr=StDev)
```
